relay/main.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	libp2p "github.com/libp2p/go-libp2p"
+	circuit "github.com/libp2p/go-libp2p-circuit"
+	inet "github.com/libp2p/go-libp2p-net"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	swarm "github.com/libp2p/go-libp2p-swarm"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+func main() {
+	// Create three libp2p hosts, enable relay client capabilities on all
+	// of them, but tell number 2 to relay connections for other peers
+	// (The ability to *use* a relay vs the ability to *be* a relay)
+	h1, err := libp2p.New(context.Background(), libp2p.EnableRelay())
+	if err != nil {
+		panic(err)
+	}
+
+	h2, err := libp2p.New(context.Background(), libp2p.EnableRelay(circuit.OptHop))
+	if err != nil {
+		panic(err)
+	}
+
+	h3, err := libp2p.New(context.Background(), libp2p.EnableRelay())
+	if err != nil {
+		panic(err)
+	}
+
+	h2info := pstore.PeerInfo{
+		ID:    h2.ID(),
+		Addrs: h2.Addrs(),
+	}
+
+	// Connect both h1 and h3 to h2, but not to eachother
+	if err := h1.Connect(context.Background(), h2info); err != nil {
+		panic(err)
+	}
+	if err := h3.Connect(context.Background(), h2info); err != nil {
+		panic(err)
+	}
+
+	// Now, to test things, let's set up a protocol handler on h3
+	h3.SetStreamHandler("/cats", func(s inet.Stream) {
+		fmt.Println("Meow! It worked!")
+		s.Close()
+	})
+
+	_, err = h1.NewStream(context.Background(), h3.ID(), "/cats")
+	if err == nil {
+		fmt.Println("Didnt actually expect to get a stream here. What happened?")
+		return
+	}
+	fmt.Println("Okay, no connection from h1 to h3: ", err)
+	fmt.Println("Just as we suspected")
+
+	// Creates a relay address
+	relayaddr, err := ma.NewMultiaddr("/p2p-circuit/ipfs/" + h3.ID().Pretty())
```

Comment:
I'd prefer to explicitly do `peer.IDB58Encode(p)`. Using `Pretty()` strings in multiaddrs seems weird, even if they're currently equivalent.